packages/firestore/src/model/document.ts

```diff
@@ -100,85 +207,68 @@ export class Document extends MaybeDocument {
   }
 
   toString(): string {
-    return (
-      `Document(${this.key}, ${this.version}, ${this.data.toString()}, ` +
-      `{hasLocalMutations: ${this.hasLocalMutations}}), ` +
-      `{hasCommittedMutations: ${this.hasCommittedMutations}})`
-    );
-  }
-
-  get hasPendingWrites(): boolean {
-    return this.hasLocalMutations || this.hasCommittedMutations;
-  }
-
-  static compareByField(field: FieldPath, d1: Document, d2: Document): number {
-    const v1 = d1.field(field);
-    const v2 = d2.field(field);
-    if (v1 !== undefined && v2 !== undefined) {
-      return v1.compareTo(v2);
-    } else {
-      return fail("Trying to compare documents on fields that don't exist");
+    switch (this.type) {
+      case DocumentType.UNKNOWN:
+        return `Document(UNKNOWN, ${this.key})`;
+      case DocumentType.CONTENTS_UNKNOWN:
+        return `Document(CONTENTS_UNKNOWN, ${this.key}, ${this.version})`;
+      case DocumentType.MISSING:
+        return `Document(MISSING, ${this.key}, ${this.version})`;
+      case DocumentType.EXISTS:
+        return (
+          `Document(${this.key}, ${this.version}, ${this.data.toString()}, ` +
+          `documentState=${this.documentState})`
+        );
+      default:
+        return fail('Unknown document type');
     }
   }
-}
-
-/**
- * A class representing a deleted document.
- * Version is set to 0 if we don't point to any specific time, otherwise it
- * denotes time we know it didn't exist at.
- */
-export class NoDocument extends MaybeDocument {
-  readonly hasCommittedMutations: boolean;
 
-  constructor(
-    key: DocumentKey,
-    version: SnapshotVersion,
-    options?: DocumentOptions
-  ) {
-    super(key, version);
-    this.hasCommittedMutations = !!(options && options.hasCommittedMutations);
+  get exists(): boolean {
+    return this.type === DocumentType.EXISTS;
   }
 
-  toString(): string {
-    return `NoDocument(${this.key}, ${this.version})`;
+  get missing(): boolean {
+    return this.type === DocumentType.MISSING;
   }
 
-  get hasPendingWrites(): boolean {
-    return this.hasCommittedMutations;
+  get unknown(): boolean {
+    return this.type === DocumentType.UNKNOWN;
   }
 
-  isEqual(other: MaybeDocument | null | undefined): boolean {
+  /**
+   * Returns true if the document is of some definite type: either existing
+   * or definitely missing.
+   */
+  get definite(): boolean {
```

Comment:
I have a really hard time guessing what `if (fooDoc.definite) { ... }` means.

In general, I'm nervous about all of these properties (`exists`, `missing`, `unknown`, `definite`) because I'm worried they could be confusing / misleading.  We're mapping 4 DocumentTypes (actually 3 since CONTENTS_UNKNOWN doesn't show up) onto 4 helper properties and the mapping isn't immediately obvious to me from the names.  You might expect that `!exists` implies `missing` but that would be wrong. 

There may be better terminology that makes everything clearer, but failing that I would be inclined to just drop them all and move the explicit `DocumentType` checks into the calling code...  if that's too onerous though, we can work harder on the names / comments.